### PR TITLE
[DTensor] Fix DTensor shardable with StridedShard

### PR DESCRIPTION
**Pull request by zpcore**
Consider shardable check for `_StridedShard`. Because the `_split_tensor` function for `_StridedShard` is to shard on `split_factor` first and then shard on tensor dim and concatenate, we need to check `split_factor` also.

### CODE REVIEW

File: test/distributed/tensor/test_utils.py [hunk 1]
```diff
@@ -1562,6 +1566,43 @@ def test_explicit_matmul(self):
...
+    def _create_spec(
+        self, mesh_shape: tuple[int, ...], placements: list[Placement]
+    ) -> DTensorSpec:
+        mesh = init_device_mesh("cpu", mesh_shape)
+        return DTensorSpec(mesh=mesh, placements=tuple(placements))
+
+    def test_is_tensor_shardable(self):
+        spec = self._create_spec((4,), [Shard(0)])
+        self.assertTrue(is_tensor_shardable([8], spec))
+        self.assertFalse(is_tensor_shardable([2], spec))
```
Comment (wconstab) [inline test/distributed/tensor/test_utils.py:1584]: nit: include the tensor shape [10] case here, should be True, unlike evenly_shardable test where it's false

File: test/distributed/tensor/test_utils.py [hunk 2]
```diff
@@ -1562,6 +1566,43 @@ def test_explicit_matmul(self):
...
+
+        spec = self._create_spec((4, 2), [Shard(0), Shard(0)])
+        self.assertTrue(is_tensor_shardable([8, 8], spec))
+
+        spec = self._create_spec((4, 2), [Shard(0), _StridedShard(0, split_factor=2)])
```
Comment (wconstab) [inline test/distributed/tensor/test_utils.py:1589]: include more cases for StridedShard? 
also let me make sure i understand the logic

1) Shard(0) on mesh dim size 4 reduces tensor shape 8 to 2
2) StridedShard(0, sf=2) on mesh dim size 2 would require 2 * 2 = 4 elts minimum? would 3 elts actually pass here or not?

File: torch/distributed/tensor/_ops/utils.py [hunk 3]
```diff
@@ -120,17 +125,21 @@ def is_tensor_shardable(
     # number of shards in each tensor dimension
     shards_map = [1] * len(shape)
     for i, placement in enumerate(spec.placements):
-        if placement.is_shard():
-            shard_dim = cast(Shard, placement).dim
+        if isinstance(placement, Shard | _StridedShard):
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/utils.py:128]: note- we have replaced one bad thing with another bad thing when we got rid of `is_shard()` helper and started copypaste Shard | _StridedShard everywhere.  I think the bare minimum is we should add a util `is_any_shard_placement` or similar.  

I might also want to introduce GenericShard base class, but I think we should decide first on the ordered-sharding plan, becuase we might want to design the inheritance structure differently if we have to have 'Strided' variants of every non-strided shard.

File: torch/distributed/tensor/_ops/utils.py [hunk 4]
```diff
@@ -120,17 +125,21 @@ def is_tensor_shardable(
...
-    for i, dim_size in enumerate(shape):
-        # TODO: maybe we should determine is_shardable based on
-        #       whether it's evenly sharded or not
-        if shards_map[i] > 1 and guard_fn(dim_size < shards_map[i]):
-            return False
+            if isinstance(placement, _StridedShard):
+                # make sure tensor dim `shard_dim` is shardable after splitting
+                # with split_factor
+                if guard_fn(
+                    shards_map[shard_dim] * placement.split_factor > shape[shard_dim]
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/utils.py]: nit: imo this would read more easily if 
1) we rename shards_map to num_shards
2) we reverse the order, shape[shard_dim] < num_shards...

**Review by wconstab — APPROVED** [review]
thanks for the fix!

File: test/distributed/tensor/test_utils.py (hunk 1 above)
Comment (zpcore) [reply test/distributed/tensor/test_utils.py:1584]: Right, added

File: torch/distributed/tensor/_ops/utils.py (hunk 3 above)
Comment (zpcore) [reply torch/distributed/tensor/_ops/utils.py:128]: Hah, yes now we almost have `Shard | _StridedShard` everywhere. `GenericShard ` may be the better choice.

File: test/distributed/tensor/test_utils.py (hunk 2 above)
Comment (zpcore) [reply test/distributed/tensor/test_utils.py:1589]: > StridedShard(0, sf=2) on mesh dim size 2 would require 2 * 2 = 4 elts minimum? would 3 elts actually pass here or not?

I think you mean 2 elts not 3 here. Right we need at least 4 elements here. The detail is:

When we apply `_StridedShard(0, split_factor=2)` on 2 elements, StridedShard will apply 2 splits: first split on split_factor, in that case we will have two parts with 1, 1 element. Second split will be based on S(0), which is not shardable as the split over 1 element over 2 parts will left a 0 element shard. Note that if the second one is shardable, we will apply concatenation back from the first split.

Below is a simple test to show if we apply `_StridedShard(0, split_factor=2)` over tensor of size 2:
```
SS = _StridedShard(0, split_factor=2)
X = torch.ones(2)
res = SS._split_tensor(X, 2)
print(res) <<<< print as ([tensor([1., 1.]), tensor([])], [0, 2])
```
So the second part is empty, which should be considered as not shardable.

File: torch/distributed/tensor/_ops/utils.py (hunk 4 above)
Comment (zpcore) [reply torch/distributed/tensor/_ops/utils.py]: Done!
